src/algebra/group_with_zero.lean

```diff
@@ -0,0 +1,504 @@
+if it is a commutative monoid with zero element. -/
+class comm_monoid_with_zero (G : Type*) extends monoid G, mul_zero_class G.
+
+/-- A type `G` is a “group with zero” if it is a monoid with zero element (distinct from `1`)
+such that every nonzero element is invertible.
+The type is required to come with an “inverse” function, and the inverse of `0` must be `0`.
+
+Examples include division rings and the ordered monoids that are the
+target of valuations in general valuation theory.-/
+class group_with_zero (G : Type*)
+  extends monoid_with_zero G, has_inv G, zero_ne_one_class G :=
+(inv_zero : (0 : G)⁻¹ = 0)
+(mul_inv_cancel : ∀ a:G, a ≠ 0 → a * a⁻¹ = 1)
+
+/-- A type `G` is a commutative “group with zero”
+if it is a commutative monoid with zero element (distinct from `1`)
+such that every nonzero element is invertible.
+The type is required to come with an “inverse” function, and the inverse of `0` must be `0`. -/
+class comm_group_with_zero (G : Type*) extends comm_monoid G, group_with_zero G.
+
+/-- The division operation on a group with zero element. -/
+instance group_with_zero.has_div {G : Type*} [group_with_zero G] :
+  has_div G := ⟨λ g h, g * h⁻¹⟩
+
+end prio
+
+lemma div_eq_mul_inv {G : Type*} [group_with_zero G] {a b : G} :
+  a / b = a * b⁻¹ := rfl
+
+namespace gwz
+variables {G : Type*} [group_with_zero G]
+
+@[simp] lemma inv_zero : (0 : G)⁻¹ = 0 :=
+group_with_zero.inv_zero G
+
+@[simp] lemma mul_inv_cancel (a : G) (h : a ≠ 0) : a * a⁻¹ = 1 :=
+group_with_zero.mul_inv_cancel a h
+
+@[simp] lemma mul_inv_cancel_assoc_left (a b : G) (h : b ≠ 0) :
+  (a * b) * b⁻¹ = a :=
+calc (a * b) * b⁻¹ = a * (b * b⁻¹) : mul_assoc _ _ _
+               ... = a             : by simp [h]
+
+@[simp] lemma mul_inv_cancel_assoc_right (a b : G) (h : a ≠ 0) :
+  a * (a⁻¹ * b) = b :=
+calc a * (a⁻¹ * b) = (a * a⁻¹) * b : (mul_assoc _ _ _).symm
+               ... = b             : by simp [h]
+
+lemma inv_ne_zero {a : G} (h : a ≠ 0) : a⁻¹ ≠ 0 :=
+assume a_eq_0, by simpa [a_eq_0] using mul_inv_cancel a h
+
+@[simp] lemma inv_mul_cancel (a : G) (h : a ≠ 0) : a⁻¹ * a = 1 :=
+calc a⁻¹ * a = (a⁻¹ * a) * a⁻¹ * a⁻¹⁻¹ : by simp [inv_ne_zero h]
+         ... = a⁻¹ * a⁻¹⁻¹             : by simp [h]
+         ... = 1                       : by simp [inv_ne_zero h]
+
+@[simp] lemma inv_mul_cancel_assoc_left (a b : G) (h : b ≠ 0) :
+  (a * b⁻¹) * b = a :=
+calc (a * b⁻¹) * b = a * (b⁻¹ * b) : mul_assoc _ _ _
+               ... = a             : by simp [h]
+
+@[simp] lemma inv_mul_cancel_assoc_right (a b : G) (h : a ≠ 0) :
+  a⁻¹ * (a * b) = b :=
+calc a⁻¹ * (a * b) = (a⁻¹ * a) * b : (mul_assoc _ _ _).symm
+               ... = b             : by simp [h]
+
+@[simp] lemma inv_inv (a : G) : a⁻¹⁻¹ = a :=
+begin
+  classical,
+  by_cases h : a = 0, { simp [h] },
+  calc a⁻¹⁻¹ = a * (a⁻¹ * a⁻¹⁻¹) : by simp [h]
+        ... = a                 : by simp [inv_ne_zero h]
+end
+
+lemma inv_involutive : function.involutive (has_inv.inv : G → G) :=
+inv_inv
+
```

Comment:
There is also an equiv you could define here I guess, inverse from G to itself.

Reply:
Yes, but I don't know how useful it is. We can always add it later.

Reply:
Agreed.